fabtests/pytest/efa/test_efa_device_selection.py

```diff
@@ -1,39 +1,54 @@
+import copy
 import pytest
-
+from efa.efa_common import efa_retrieve_hw_counter_value, get_efa_device_names
+from common import ClientServerTest
 
 # This test must be run in serial mode because it checks the hw counter
 @pytest.mark.serial
 @pytest.mark.functional
 def test_efa_device_selection(cmdline_args):
-    from efa.efa_common import efa_retrieve_hw_counter_value, get_efa_domain_names
-    from common import ClientServerTest
 
     if cmdline_args.server_id == cmdline_args.client_id:
         pytest.skip("EFA device selection test requires 2 nodes")
         return
 
-    efa_domain_names = get_efa_domain_names(cmdline_args.server_id)
+    server_device_names = get_efa_device_names(cmdline_args.server_id)
+    client_device_names = get_efa_device_names(cmdline_args.client_id)
+
+    server_num_devices = len(server_device_names)
+    client_num_devices = len(server_device_names)
+
+    for i in range(max(server_num_devices, client_num_devices)):
+
+        server_device_idx = i % server_num_devices
+        client_device_idx = i % client_num_devices
+
+        server_device_name = server_device_names[server_device_idx]
+        client_device_name = client_device_names[client_device_idx]
+
+        for suffix in ["rdm", "dgrm"]:
+
+            server_tx_bytes_before_test = efa_retrieve_hw_counter_value(cmdline_args.server_id, "tx_bytes", server_device_name)
+            client_tx_bytes_before_test = efa_retrieve_hw_counter_value(cmdline_args.client_id, "tx_bytes", client_device_name)
 
-    for efa_domain_name in efa_domain_names:
-        if '-rdm' in efa_domain_name:
-            assert not('-dgrm') in efa_domain_name
-            fabtest_opts = "fi_rdm_pingpong"
-        elif '-dgrm' in efa_domain_name:
-            assert not('-rdm') in efa_domain_name
-            fabtest_opts = "fi_dgram_pingpong -k"
+            if suffix == "rdm":
+                command = "fi_rdm_pingpong"
+            else:
+                command = "fi_dgram_pingpong -k" # efa provider requires prefix mode for dgram provider, hence "-k"
 
-        efa_device_name = efa_domain_name.split('-')[0]
+            server_domain_name = server_device_name + "-" + suffix
+            client_domain_name = client_device_name + "-" + suffix
 
-        server_tx_bytes_before_test = efa_retrieve_hw_counter_value(cmdline_args.server_id, "tx_bytes", efa_device_name)
-        client_tx_bytes_before_test = efa_retrieve_hw_counter_value(cmdline_args.client_id, "tx_bytes", efa_device_name)
+            cmdline_args_copy = copy.copy(cmdline_args)
```

Comment:
`copy.copy` does a shallow copy. If you want to keep the original `cmdline_args` unchanged, you need to use `copy.deepcopy`

Reply:
If `additional_server_arguments` and `additional_client_arguments` are not set in the original `cmdline_args`, then a shallow copy is good enough

Reply:
I think actually no copy is required at all (as long as ClientServerTest doesn't modify it) because each loop we just assign a new value to additional_*_arguments.

Reply:
I think copy is needed because cmdline_args is a fixture created once during initialization.

I think we can change the scope of `cmdline_args`. 

@sunkuamzn  I tried to use deepcopy, but it failed with an error message about not able to deepcopy a pattern object?
Since copy is fine here, I will keep this

Reply:
👍